### PR TITLE
feat(kafka-logger): supports logging request body

### LINKED ISSUE #5343: request help: kafka logger supports logging request body
### Issue description

Add a request_body switch to the schema, and each body can be used by expr to decide whether to log or not. Without this switch, the body is not logged.

```
"kafka-logger": {
   "broker_list":{
       "127.0.0.1":9092
    },
   "kafka_topic" : "test2",
   "request_body": {
       "expr": [
          ["request_length", "<", "1024"],
       ]
   },
   "key" : "key1",
   "batch_max_size": 1,
   "name": "kafka logger"
}
```

`expr` can be evaluated by lua-resty-expr. request body can be fetched by core.request.get_body.

### Environment

- apisix version (cmd: `apisix version`): 2.10.1
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

### PR DESCRIPTION
**Pull request by windyrjc**
### What this PR does / why we need it:
kafka logger supports logging request body 
fix #5343 
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/kafka-logger.lua [hunk 1]
```diff
@@ -74,6 +74,15 @@ local schema = {
         inactive_timeout = {type = "integer", minimum = 1, default = 5},
         batch_max_size = {type = "integer", minimum = 1, default = 1000},
         include_req_body = {type = "boolean", default = false},
+        request_body_expr = {
```
Comment (spacewander) [inline apisix/plugins/kafka-logger.lua:77]: Would be better to name it as `include_req_body_expr`?

File: apisix/utils/log-util.lua [hunk 2]
```diff
@@ -119,13 +120,32 @@ local function get_full_log(ngx, conf)
     }
 
     if conf.include_req_body then
```
Comment (spacewander) [inline apisix/utils/log-util.lua:122]: I think we can change the current logic to
```
if include_req_body || eval(include_req_body_expr) {
  log req body
}
```
What is your opinion?

File: t/plugin/kafka-logger.t [hunk 3]
```diff
@@ -1114,3 +1114,98 @@ GET /t
...
+POST /hello
+abcdef
+--- response_body
+hello world
+--- no_error_log
+[error]
+--- error_log_like eval
+qr/send data to kafka: \{.*"body":"abcdef"/
+--- wait: 2
+--- wait: 2
```
Comment (spacewander) [inline t/plugin/kafka-logger.t:1211]: Duplicate `--- wait`?

File: apisix/utils/log-util.lua [hunk 4]
```diff
@@ -119,13 +120,32 @@ local function get_full_log(ngx, conf)
...
-            log.request.body = body
-        else
-            local body_file = ngx.req.get_body_file()
-            if body_file then
-                log.request.body_file = body_file
+
+        local log_request_body = true
+
+        if conf.request_body_expr then
+            local request_expr, err = expr.new(conf.request_body_expr)
```
Comment (spacewander) [inline apisix/utils/log-util.lua:127]: We can check `conf.request_expr` before new?

File: apisix/utils/log-util.lua [hunk 5]
```diff
@@ -119,13 +120,32 @@ local function get_full_log(ngx, conf)
...
+            if not request_expr then
+                core.log.error('generate log expr err ' .. err)
```
Comment (spacewander) [inline apisix/utils/log-util.lua:129]: Need to skip the remaining logic. We should not use a nil request_expr.

File: apisix/utils/log-util.lua (hunk 2 above)
Comment (windyrjc) [reply apisix/utils/log-util.lua:122]: In my opinion, i think `include_req_body` like a main switch, `include_req_body_expr` like a switch in condition. Close  `include_req_body`  will not log any request body data, and `include_req_body`'s semantic is compatible with  the previous version of kafka-logger, we are just only add a condition expr.

File: apisix/plugins/kafka-logger.lua (hunk 1 above)
Comment (windyrjc) [reply apisix/plugins/kafka-logger.lua]: done

File: t/plugin/kafka-logger.t (hunk 3 above)
Comment (windyrjc) [reply t/plugin/kafka-logger.t]: done

File: apisix/utils/log-util.lua (hunk 4 above)
Comment (windyrjc) [reply apisix/utils/log-util.lua]: done

File: apisix/utils/log-util.lua (hunk 5 above)
Comment (windyrjc) [reply apisix/utils/log-util.lua]: done

File: t/plugin/kafka-logger.t [hunk 6]
```diff
@@ -1114,3 +1114,97 @@ GET /t
...
+                                  {
+                                    "127.0.0.1":9092
+                                  },
+                                "kafka_topic" : "test2",
+                                "key" : "key1",
+                                "timeout" : 1,
+                                "include_req_body": true,
+                                "request_body_expr": [
+                                    [
+                                      "remote_addr",
```
Comment (spacewander) [inline t/plugin/kafka-logger.t:1140]: Better to use a test case that can eval to false.

File: t/plugin/kafka-logger.t [hunk 7]
```diff
@@ -1114,3 +1114,97 @@ GET /t
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 27: hit route, report log to kafka
```
Comment (spacewander) [inline t/plugin/kafka-logger.t:1200]: Let's add a test that expr is evaluated to false

File: docs/en/latest/plugins/kafka-logger.md [hunk 8]
```diff
@@ -58,6 +58,7 @@ For more info on Batch-Processor in Apache APISIX please refer.
 | retry_delay      | integer | optional    | 1              | [0,...] | Number of seconds the process execution should be delayed if the execution fails.        |
 | include_req_body | boolean | optional    | false          | [false, true] | Whether to include the request body. false: indicates that the requested body is not included; true: indicates that the requested body is included. |
 | cluster_name     | integer | optional    | 1              | [0,...] | the name of the cluster. When there are two or more kafka clusters, you can specify different names. And this only works with async producer_type.|
+| request_body_expr  | object  | optional    |                |         | Whether to logging request body,based on [lua-resty-expr](https://github.com/api7/lua-resty-expr).            |
```
Comment (spacewander) [inline docs/en/latest/plugins/kafka-logger.md:61]: Better to move it under include_req_body and explain its relation to include_req_body.

Comment (windyrjc) [reply docs/en/latest/plugins/kafka-logger.md]: done

File: t/plugin/kafka-logger.t (hunk 7 above)
Comment (windyrjc) [reply t/plugin/kafka-logger.t:1200]: done

File: t/plugin/kafka-logger.t (hunk 6 above)
Comment (windyrjc) [reply t/plugin/kafka-logger.t:1140]: done

File: t/plugin/kafka-logger.t [hunk 9]
```diff
@@ -1135,7 +1135,7 @@ qr/failed to send data to Kafka topic: .*, brokers: \{"127.0.0.127":9092}/
                                 "key" : "key1",
                                 "timeout" : 1,
                                 "include_req_body": true,
-                                "request_body_expr": [
+                                "include_req_body_expr": [
                                     [
                                       "remote_addr",
```
Comment (spacewander) [inline t/plugin/kafka-logger.t:1140]: We can do it with `arg_xx`. There is no need to use a separate rule.

File: t/plugin/kafka-logger.t [hunk 10]
```diff
@@ -1208,3 +1208,107 @@ hello world
 --- error_log_like eval
 qr/send data to kafka: \{.*"body":"abcdef"/
 --- wait: 2
+
+
+
+=== TEST 28: hit route, not trigger request_body_expr rule
+--- request
+GET /hello
```
Comment (spacewander) [inline t/plugin/kafka-logger.t:1216]: The test will pass whether the request_body_expr is correct or not

File: t/plugin/kafka-logger.t [hunk 11]
```diff
@@ -1114,3 +1114,201 @@ GET /t
...
+                        },
+                        "upstream": {
+                            "nodes": {
+                                "127.0.0.1:1980": 1
+                            },
+                            "type": "roundrobin"
+                        },
+                        "uri": "/hello"
+                }]],
+                [[{
```
Comment (spacewander) [inline t/plugin/kafka-logger.t:1156]: Please remove the part which is not relative to the test, from L1156 to L1182

File: t/plugin/kafka-logger.t (hunk 10 above)
Comment (windyrjc) [reply t/plugin/kafka-logger.t:1216]: Ok, i just want to prove that if there was no request body, it still work normally. There is already has a eval false case below and i'm delete this case for good

File: t/plugin/kafka-logger.t (hunk 9 above)
Comment (windyrjc) [reply t/plugin/kafka-logger.t:1140]: I'm not understand what this means

Comment (spacewander) [reply t/plugin/kafka-logger.t:1140]: Use `arg_blah == 1` in the expression, and hit it with "POST /hello?blah=1" and "POST /hello?blah=2".

Comment (windyrjc) [reply t/plugin/kafka-logger.t:1140]: > Use `arg_blah == 1` in the expression, and hit it with "POST /hello?blah=1" and "POST /hello?blah=2".

I'm using `ctx.var` to eval the expr and  `ctx.var` don't have request param, it only has request urI.If you want to achieve this, i need to eval the expr by `ngx.req.get_uri_args`

Comment (windyrjc) [reply t/plugin/kafka-logger.t:1140]: I'm using ctx.var to eval the expr and ctx.var doesn't directly hold args. it only has args table. If you want to achieve this, i need to eval the expr by ctx.var.args

Comment (spacewander) [reply t/plugin/kafka-logger.t:1140]: I mean to use `$arg_` variable. You can refer to https://github.com/apache/apisix/blob/cc43b9fc1bf9be14c05d36415f83cdd189d0a7f5/t/plugin/traffic-split4.t#L110

Comment (windyrjc) [reply t/plugin/kafka-logger.t]: > I mean to use `$arg_` variable. You can refer to
> 
> https://github.com/apache/apisix/blob/cc43b9fc1bf9be14c05d36415f83cdd189d0a7f5/t/plugin/traffic-split4.t#L110

done ,thank you very much

File: t/plugin/kafka-logger.t (hunk 11 above)
Comment (windyrjc) [reply t/plugin/kafka-logger.t]: done

File: docs/zh/latest/plugins/kafka-logger.md [hunk 12]
```diff
@@ -57,6 +57,7 @@ title: kafka-logger
 | max_retry_count  | integer | 可选   | 0              | [0,...] | 从处理管道中移除之前的最大重试次数。             |
 | retry_delay      | integer | 可选   | 1              | [0,...] | 如果执行失败，则应延迟执行流程的秒数。           |
 | include_req_body | boolean | 可选   | false          | [false, true] | 是否包括请求 body。false： 表示不包含请求的 body ； true： 表示包含请求的 body 。|
+| include_req_body_expr | array  | 可选    |           |         | 是否采集请求body,基于[lua-resty-expr](https://github.com/api7/lua-resty-expr)。 该选项需要开启 `include_req_body`|
```
Comment (tzssangglass) [inline docs/zh/latest/plugins/kafka-logger.md:60]:
```suggestion
| include_req_body_expr | array  | 可选    |           |         | 是否采集请求body，基于 [lua-resty-expr](https://github.com/api7/lua-resty-expr)。 该选项需要开启 `include_req_body`|
```

File: docs/en/latest/plugins/kafka-logger.md [hunk 13]
```diff
@@ -57,6 +57,7 @@ For more info on Batch-Processor in Apache APISIX please refer.
 | max_retry_count  | integer | optional    | 0              | [0,...] | Maximum number of retries before removing from the processing pipe line.                 |
 | retry_delay      | integer | optional    | 1              | [0,...] | Number of seconds the process execution should be delayed if the execution fails.        |
 | include_req_body | boolean | optional    | false          | [false, true] | Whether to include the request body. false: indicates that the requested body is not included; true: indicates that the requested body is included. |
+| include_req_body_expr  | array  | optional    |          |         | Whether to logging request body,based on [lua-resty-expr](https://github.com/api7/lua-resty-expr), this option require to turn on `include_req_body` option.             |
```
Comment (tzssangglass) [inline docs/en/latest/plugins/kafka-logger.md:60]:
```suggestion
| include_req_body_expr  | array  | optional    |          |         | Whether to logging request body, based on [lua-resty-expr](https://github.com/api7/lua-resty-expr), this option require to turn on `include_req_body` option.             |
```

File: t/plugin/kafka-logger.t [hunk 14]
```diff
@@ -1114,3 +1114,80 @@ GET /t
...
+--- response_body
+hello world
+--- no_error_log
+[error]
+--- error_log eval
+qr/send data to kafka: \{.*"body":"abcdef"/
+--- wait: 2
+
+
+=== TEST 28: hit route,expr eval fail
```
Comment (tzssangglass) [inline t/plugin/kafka-logger.t:1185]:
```suggestion
--- wait: 2



=== TEST 28: hit route,expr eval fail
```

File: t/plugin/kafka-logger.t [hunk 15]
```diff
@@ -1114,3 +1114,80 @@ GET /t
...
+
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+=== TEST 27: hit route, expr eval success
```
Comment (tzssangglass) [inline t/plugin/kafka-logger.t:1172]:
```suggestion
[error]



=== TEST 27: hit route, expr eval success
```

Comment (windyrjc) [reply t/plugin/kafka-logger.t]: i forgot lint before push, sorry~

File: apisix/plugins/kafka-logger.lua [hunk 16]
```diff
@@ -74,6 +74,15 @@ local schema = {
...
         include_req_body = {type = "boolean", default = false},
+        include_req_body_expr = {
+            type = "array",
+            items = {
+                type = "array",
+                items = {
+                    type = "string"
+                }
+            }
+        },
```
Comment (tzssangglass) [inline apisix/plugins/kafka-logger.lua:85]:
```suggestion
        include_req_body_expr = {
            type = "array",
            minItems = 1,
            items = {
                type = "array",
                items = {
                    type = "string"
                }
            }
        },
```

File: t/plugin/kafka-logger.t [hunk 17]
```diff
@@ -1114,3 +1114,82 @@ GET /t
...
+
+
+=== TEST 26: set route(id: 1,include_req_body = true,include_req_body_expr = array)
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
```
Comment (tzssangglass) [inline t/plugin/kafka-logger.t:1127]:
```suggestion
                 [=[{
```

File: t/plugin/kafka-logger.t [hunk 18]
```diff
@@ -1114,3 +1114,82 @@ GET /t
...
+                            }
+                        },
+                        "upstream": {
+                            "nodes": {
+                                "127.0.0.1:1980": 1
+                            },
+                            "type": "roundrobin"
+                        },
+                        "uri": "/hello"
+                }]]
```
Comment (tzssangglass) [inline t/plugin/kafka-logger.t:1155]:
```suggestion
                }]=]
```

File: apisix/plugins/kafka-logger.lua [hunk 19]
```diff
@@ -74,6 +74,15 @@ local schema = {
         inactive_timeout = {type = "integer", minimum = 1, default = 5},
         batch_max_size = {type = "integer", minimum = 1, default = 1000},
         include_req_body = {type = "boolean", default = false},
+        include_req_body_expr = {
```
Comment (spacewander) [inline apisix/plugins/kafka-logger.lua:77]: Before we can merge it, could you add a check for the expr in the `check_schema`? Like:
https://github.com/apache/apisix/blob/9b355c6f13194c00eeb491542ee0727467988877/apisix/plugins/response-rewrite.lua#L111

**Review by spacewander — APPROVED** [review]

File: apisix/plugins/kafka-logger.lua (hunk 19 above)
Comment (windyrjc) [reply apisix/plugins/kafka-logger.lua:78]: done

**Review by tzssangglass — APPROVED** [review]

File: apisix/plugins/kafka-logger.lua [hunk 20]
```diff
@@ -107,6 +109,14 @@ local _M = {
 
 
 function _M.check_schema(conf, schema_type)
+
+    if conf.vars then
```
Comment (spacewander) [inline apisix/plugins/kafka-logger.lua:113]: Should be `include_req_body_expr `

**Review by spacewander — APPROVED** [review]

**Pull request merged**